juju/juju
# K8s proxy support4register controller2.9

back-port https://github.com/juju/juju/pull/14878
 
## QA steps

```sh
$ juju add-user admin1
User "admin1" added
Please send this command to admin1:
    juju register MIIEtxMGYWRtaW4xMBUTEzEwLjE1Mi4xODMuODg6MTcwNzAEIAkwsKvHf7cXHwFLgLZ_JLoh12Z5LrDHYdkPLauB-E1rEwJrMQyCBG50eXBlOiBrdWJlcm5ldGVzLXBvcnQtZm9yd2FyZApjb25maWc6CiAgYXBpLWhvc3Q6IGh0dHBzOi8vMTkyLjE2OC4wLjE2OToxNjQ0MwogIGNhLWNlcnQ6ICIiCiAgbmFtZXNwYWNlOiBjb250cm9sbGVyLWsxCiAgcmVtb3RlLXBvcnQ6ICIxNzA3MCIKICBzZXJ2aWNlOiBjb250cm9sbGVyLXNlcnZpY2UKICBzZXJ2aWNlLWFjY291bnQtdG9rZW46IGV5SmhiR2NpT2lKU1V6STFOaUlzSW10cFpDSTZJa3htZEhaNmIwczFkMEkxVDFsT1IzZEZXbmhhTm5WcGFWaDVhak5OYW5KTlltMU1OVmt0UVdRNGRtOGlmUS5leUpwYzNNaU9pSnJkV0psY201bGRHVnpMM05sY25acFkyVmhZMk52ZFc1MElpd2lhM1ZpWlhKdVpYUmxjeTVwYnk5elpYSjJhV05sWVdOamIzVnVkQzl1WVcxbGMzQmhZMlVpT2lKamIyNTBjbTlzYkdWeUxXc3hJaXdpYTNWaVpYSnVaWFJsY3k1cGJ5OXpaWEoyYVdObFlXTmpiM1Z1ZEM5elpXTnlaWFF1Ym1GdFpTSTZJbU52Ym5SeWIyeHNaWEl0Y0hKdmVIa2lMQ0pyZFdKbGNtNWxkR1Z6TG1sdkwzTmxjblpwWTJWaFkyTnZkVzUwTDNObGNuWnBZMlV0WVdOamIzVnVkQzV1WVcxbElqb2lZMjl1ZEhKdmJHeGxjaTF3Y205NGVTSXNJbXQxWW1WeWJtVjBaWE11YVc4dmMyVnlkbWxqWldGalkyOTFiblF2YzJWeWRtbGpaUzFoWTJOdmRXNTBMblZwWkNJNkltRmtaR0UzTVRCbUxUYzNOall0TkRreU1DMWlNR00wTFdaak1EZzBORGs0TkRZNVlTSXNJbk4xWWlJNkluTjVjM1JsYlRwelpYSjJhV05sWVdOamIzVnVkRHBqYjI1MGNtOXNiR1Z5TFdzeE9tTnZiblJ5YjJ4c1pYSXRjSEp2ZUhraWZRLkxLakFfWmpxeXNFN19ubms3cWZBRDVUTm9aWW1ZYmRITm9sUGt1YWRLNWw1X1BDeUZjd1NXc2c0NGlMVnlhWUM0a0NtVFkwckNPalZfTFRMZXd1MXNDQzFfTzFlbjRpSWx5SGpiOFdUdE1sVExXTXJvZHlmSThoLXpuNVQyVGxFQ1FueHZJSndEUTd3NTJselZvbmdSSjVvY2JNM2tTUGZFLW83UkotT3VBZ0NwQUdsbE1wWXhNMzdVcm90eDA1enlhenRfN3VFVThJUndoLVNOZHRRZHJ5cVltRUVlNkEwcmc5SzhGU2R1X2I3MlNjenM2RU9zV3JLb3plSVVUY3ZLT1Y1Q0JfVkZmM3hMTjkwalJfQTdxaHFhbmR4OGt4R3BYcWJ2MVNhc0g0NzA2dkN5QjRSQ3pKX2pQMEk5OVdNLVBmY1lRRjBNQzJqZngyM0xfSC0yQQoA

"admin1" has not been granted access to any models. You can use "juju grant" to grant access.

$ juju add-model t1
Added 't1' model on microk8s/localhost with credential 'microk8s' for user 'admin'

$ juju grant admin1 admin t1

$ export JUJU_DATA=/tmp/test-juju

$ juju models
Since Juju 2 is being run for the first time, it has downloaded the latest public cloud information.
No controllers registered.

Please either create a new controller using "juju bootstrap" or connect to
another controller that you have been given access to using "juju register".
ERROR No controllers registered.

Please either create a new controller using "juju bootstrap" or connect to
another controller that you have been given access to using "juju register".

$ juju register MIIEtxMGYWRtaW4xMBUTEzEwLjE1Mi4xODMuODg6MTcwNzAEIAkwsKvHf7cXHwFLgLZ_JLoh12Z5LrDHYdkPLauB-E1rEwJrMQyCBG50eXBlOiBrdWJlcm5ldGVzLXBvcnQtZm9yd2FyZApjb25maWc6CiAgYXBpLWhvc3Q6IGh0dHBzOi8vMTkyLjE2OC4wLjE2OToxNjQ0MwogIGNhLWNlcnQ6ICIiCiAgbmFtZXNwYWNlOiBjb250cm9sbGVyLWsxCiAgcmVtb3RlLXBvcnQ6ICIxNzA3MCIKICBzZXJ2aWNlOiBjb250cm9sbGVyLXNlcnZpY2UKICBzZXJ2aWNlLWFjY291bnQtdG9rZW46IGV5SmhiR2NpT2lKU1V6STFOaUlzSW10cFpDSTZJa3htZEhaNmIwczFkMEkxVDFsT1IzZEZXbmhhTm5WcGFWaDVhak5OYW5KTlltMU1OVmt0UVdRNGRtOGlmUS5leUpwYzNNaU9pSnJkV0psY201bGRHVnpMM05sY25acFkyVmhZMk52ZFc1MElpd2lhM1ZpWlhKdVpYUmxjeTVwYnk5elpYSjJhV05sWVdOamIzVnVkQzl1WVcxbGMzQmhZMlVpT2lKamIyNTBjbTlzYkdWeUxXc3hJaXdpYTNWaVpYSnVaWFJsY3k1cGJ5OXpaWEoyYVdObFlXTmpiM1Z1ZEM5elpXTnlaWFF1Ym1GdFpTSTZJbU52Ym5SeWIyeHNaWEl0Y0hKdmVIa2lMQ0pyZFdKbGNtNWxkR1Z6TG1sdkwzTmxjblpwWTJWaFkyTnZkVzUwTDNObGNuWnBZMlV0WVdOamIzVnVkQzV1WVcxbElqb2lZMjl1ZEhKdmJHeGxjaTF3Y205NGVTSXNJbXQxWW1WeWJtVjBaWE11YVc4dmMyVnlkbWxqWldGalkyOTFiblF2YzJWeWRtbGpaUzFoWTJOdmRXNTBMblZwWkNJNkltRmtaR0UzTVRCbUxUYzNOall0TkRreU1DMWlNR00wTFdaak1EZzBORGs0TkRZNVlTSXNJbk4xWWlJNkluTjVjM1JsYlRwelpYSjJhV05sWVdOamIzVnVkRHBqYjI1MGNtOXNiR1Z5TFdzeE9tTnZiblJ5YjJ4c1pYSXRjSEp2ZUhraWZRLkxLakFfWmpxeXNFN19ubms3cWZBRDVUTm9aWW1ZYmRITm9sUGt1YWRLNWw1X1BDeUZjd1NXc2c0NGlMVnlhWUM0a0NtVFkwckNPalZfTFRMZXd1MXNDQzFfTzFlbjRpSWx5SGpiOFdUdE1sVExXTXJvZHlmSThoLXpuNVQyVGxFQ1FueHZJSndEUTd3NTJselZvbmdSSjVvY2JNM2tTUGZFLW83UkotT3VBZ0NwQUdsbE1wWXhNMzdVcm90eDA1enlhenRfN3VFVThJUndoLVNOZHRRZHJ5cVltRUVlNkEwcmc5SzhGU2R1X2I3MlNjenM2RU9zV3JLb3plSVVUY3ZLT1Y1Q0JfVkZmM3hMTjkwalJfQTdxaHFhbmR4OGt4R3BYcWJ2MVNhc0g0NzA2dkN5QjRSQ3pKX2pQMEk5OVdNLVBmY1lRRjBNQzJqZngyM0xfSC0yQQoA
Enter a new password:
Confirm password:
Enter a name for this controller [k1]:
Initial password successfully set for admin1.

Welcome, admin1. You are now logged into "k1".

$ cat $JUJU_DATA/controllers.yaml | yq .controllers.k1."proxy-config"
type: kubernetes-port-forward
config:
  api-host: https://10.152.183.1:443
  ca-cert: |
    -----BEGIN CERTIFICATE-----
    ...
    E3nyriHIR7oE4iZPHfS0aeEouw==
    -----END CERTIFICATE-----
  namespace: controller-k1
  remote-port: "17070"
  service: controller-service
  service-account-token: eyJhbGciOiJ..._H-2A

$ juju models
Controller: k1

Model     Cloud/Region        Type        Status     Access  Last connection
admin/t1  microk8s/localhost  kubernetes  available  admin   never connected

$ juju switch k1:admin/t1
k1 (controller) -> k1:admin/t1

$ juju deploy hello-kubecon
Located charm "hello-kubecon" in charm-hub, revision 15
Deploying "hello-kubecon" from charm-hub charm "hello-kubecon", revision 15 in channel stable on focal

$ juju status
Model  Controller  Cloud/Region        Version  SLA          Timestamp
t1     k1          microk8s/localhost  2.9.38   unsupported  15:34:03+11:00

App            Version  Status  Scale  Charm          Channel  Rev  Address         Exposed  Message
hello-kubecon           active      1  hello-kubecon  stable    15  10.152.183.216  no

Unit              Workload  Agent  Address      Ports  Message
hello-kubecon/0*  active    idle   10.1.97.202
```
